File: crates/taflow-core/src/indicators/butterfly_pattern.rs
use crate::error::TaResult;
use super::pattern_swing::{ratios_in, xabcd, SwingTracker, SWING_THRESHOLD};

/// Butterfly X-A-B-C-D harmonic completion signal.
#[derive(Debug, Clone)]
pub struct ButterflyPattern { swing: SwingTracker, count: usize, value: Option<f64> }
impl ButterflyPattern {
    /// Create a detector retaining five confirmed pivots.
    pub fn new()->TaResult<Self>{Ok(Self{swing:SwingTracker::new(SWING_THRESHOLD,5),count:0,value:None})}
    /// Append one OHLC bar and return the latest harmonic signal.
    pub fn append(&mut self,_open:f64,high:f64,low:f64,_close:f64)->Option<f64>{self.count+=1;self.value=Some(0.0);if !self.swing.append(high,low)||self.swing.pivots().len()<5{return self.value;}let p=xabcd(self.swing.pivots());let xa=(p.a-p.x).abs();let ab=(p.b-p.a).abs();let bc=(p.c-p.b).abs();let cd=(p.d-p.c).abs();let ad=(p.d-p.a).abs();if ratios_in(&[(ab/xa,.74,.84),(bc/ab,.382,.886),(cd/bc,1.618,2.618),(ad/xa,1.27,1.618)]){self.value=Some(if p.bullish{1.0}else{-1.0});}self.value}
    /// Return the latest signal.
    pub fn value(&self)->Option<f64>{self.value}
    /// Return the processed-bar count.
    pub fn len(&self)->usize{self.count}
    /// Return whether no bars were processed.
    pub fn is_empty(&self)->bool{self.count==0}
    /// Clear pivots and output.
    pub fn reset(&mut self){self.swing.reset();self.count=0;self.value=None;}
}

File: python/taflow/indicators/butterfly_pattern.py
from typing import Any

import numpy as np

from .._native import ButterflyPattern as _Native
from .._series import as_float64_series


class ButterflyPattern:
    """Causal Butterfly harmonic reversal detector over aligned OHLC bars.

    Rust owns the bounded persistent window, pattern arithmetic, warm-up, and
    output history. The required input order is open, high, low, close. Scalar
    output is 1.0 for a bullish completion, -1.0 for a bearish completion, and
    0.0 otherwise, including before five confirmed alternating pivots exist.
    Wickra's XA, AB, BC, CD, and AD ratio ranges are evaluated only when a new
    pivot is confirmed, so historical signals are causal and do not repaint.
    append, extend, and reset mutate and return this adapter. The external name
    mapping is Wickra 0.9.9 Butterfly; TA-Lib has no equivalent pattern function.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.

        Empty aligned series create a fresh streaming state. Differing input
        lengths raise ValueError before native state is mutated.
        """
        self._state = _Native()

    def append(
        self, open: float, high: float, low: float, close: float
    ) -> "ButterflyPattern":
        """Append one open/high/low/close bar and return this adapter."""
        self._state.append(float(open), float(high), float(low), float(close))
        return self

    def extend(self, open: Any, high: Any, low: Any, close: Any) -> "ButterflyPattern":
        """Append aligned OHLC histories and return this adapter."""
        series = tuple(as_float64_series(item) for item in (open, high, low, close))
        if len({len(item) for item in series}) != 1:
            raise ValueError("OHLC inputs must have equal lengths")
        self._state.extend(*series)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest native signal, or None during warm-up."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned native signal history as a NumPy array."""
        return self._state.compute()

    def reset(self) -> "ButterflyPattern":
        """Restore fresh native state without reallocating and return self."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)
